### ISSUE TITLE
Working on two tex files in different GUI-windows: inverse search messes the window setup up

### ISSUE THREAD

**Issue by kiryph** [lervag/vimtex#546]
I am on OSX using MacVim.
I have opened two GUI-windows of MacVim. Each of them contains a different main tex file.
I have opened both PDF's in different GUI-windows of pdf-viewer skim.
Forward search works as expected.

The issue is the inverse search: it always uses the same MacVim Window:

```
Note: Inverse search relies on the |clientserver| functionality of vim.  Each
      instance of vim runs its own server.  The above mentioned standard value
      implies the default server name, which might not be the actual name of
      the server for the vim instance you are using.
```

Afaik, this is a principle issue due to the settings in the pdf viewer. They try to connect always to the same vim server.

Is there any possibility to circumvent this problem?

**Comment by kiryph** [comment]
My current search has not been fruitful except for two ideas:

- extend https://github.com/susam/vimer to handle multiple servers if they exist
- extend http://skim-app.sourceforge.net/ to accept a new command line option of displayline which adjusts temporarily SKTeXEditorArguments for a Skim instance when doing a forward search.

**Comment by lervag** [comment]
Is `zathura` available for MacOS? If it is, then you could consider using that instead. It allows Vim to set the editor directly, which makes it possible to supply the vim server.

For other viewers, I can think of some ways to handle this, but I have not implemented any. Mostly because I don't need them myself, and because they are hackish. But you might want to try yourself. So: One thing that could work is to make Vim/Vimtex save to either a file or an environmental variable every project that is open and the corresponding vim server. This could be done by hooking into the events, i.e. `:h vimtex-events`. The InitPost event can be used to e.g. write to file the path to the current tex project main file, as well as the vim server, and the Quit event could remove the entry. Then, from the viewer side, you could use a bash script that parsed the list of projects and choose the correct vim server based on the file opened in the viewer. The backward search settings in most viewers allow to send the tex file name, so I think this might work.

But: I don't have the time to try this myself... feel free to open a pull request if you find a way to do it that seems general, or of you have a method that could be documented, e.g. in the FAQ section.

**Comment by kiryph** [comment]
No, `zathura` is not available for macOS. Thanks for your thoughts. But I think the right direction is the following even though I could not resolve it entirely:

Application defaults on macOS are written to so called property lists which can be checked via

```
$ defaults read $HOME/Library/Preferences/net.sourceforge.skim-app.skim.plist | grep TeX
    SKTeXEditorArguments = "--remote-silent +\\"%line\\" \\"%file\\"";
    SKTeXEditorCommand = mvim;
    SKTeXEditorPreset = MacVim;
```

These can be overwritten via command line arguments when a new Skim application is launched
http://perspx.com/archives/parsing-command-line-arguments-nsuserdefaults/

For example:
```
$ /Applications/Skim.app/Contents/MacOS/Skim \
 -SKTeXEditorPreset "Custom" \
 -SKTeXEditorCommand "mvim" \
 -SKTeXEditorArguments '"--remote-silent +\\"%line\\" \\"%file\\" --server VIM1"' \
 $HOME/Manuscripts/thesis.pdf
```

However, the recommended way to open a pdf file in Skim when using synctex is to use the bash script `displayline` of `Skim.app` which does not necessarily opens a new Skim instance.
https://sourceforge.net/p/skim-app/wiki/TeX_and_PDF_Synchronization/#setting-up-your-editor-for-forward-search

Ths bash script `displayline` looks like the following and contains an here-doc section of AppleScript with a few specific options:
```
$ cat /Applications/Skim.app/Contents/SharedSupport/displayline
#!/bin/bash

# displayline (Skim)
#
# Usage: displayline [-r] [-b] [-g] LINE PDFFILE [TEXSOURCEFILE]

if [[ $# -eq 0 || "$1" == "-h" || "$1" == "-help" ]]; then
  echo "Usage: displayline [-r] [-b] [-g] LINE PDFFILE [TEXSOURCEFILE]
Options:
-r, -revert      Revert the file from disk if it was open
-b, -readingbar  Indicate the line using the reading bar
-g, -background  Do not bring Skim to the foreground"
  exit 0
fi

# get arguments
revert=false
bar=false
activate=true
while [[ "${1:0:1}" == "-" ]]; do
  if [[ "$1" == "-r" || "$1" == "-revert" ]]; then
    revert=true
  elif [[ "$1" == "-b" || "$1" == "-readingbar" ]]; then
    bar=true
  elif [[ "$1" == "-g" || "$1" == "-background" ]]; then
    activate=false
  fi
  shift
done
line=$1
file="$2"
shopt -s extglob
[[ $# -gt 2 ]] && source="$3" || source="${file%.@(pdf|dvi|xdv)}.tex"

# expand relative paths
[[ "${file:0:1}" == "/" ]] || file="${PWD}/${file}"
[[ "${source:0:1}" == "/" ]] || source="${PWD}/${source}"

# run AppleScript
/usr/bin/osascript << EOF
  set theLine to $line as integer
  set theFile to POSIX file "$file"
  set theSource to POSIX file "$source"
  set thePath to POSIX path of (theFile as alias)
  tell application "Skim"
    if $activate then activate
    if $revert then
      try
        set theDocs to get documents whose path is thePath
        if (count of theDocs) > 0 then revert theDocs
      end try
    end if
    open theFile
    tell front document to go to TeX line theLine from theSource showing reading bar $bar
  end tell
EOF
```

The features of Skim which can be accessed via AppleScript are documented here:
https://sourceforge.net/p/skim-app/wiki/AppleScript/

Unfortunately, the preferences are not listed. So I am back at square 1.

**Comment by lervag** [comment]
Thanks for the info. It seems you are on the right track. I'm sorry that I can't be of any more help regarding this (I don't own or use any Macs).

Let me know if you find any solution that can or should be documented.

**Comment by kiryph** [comment]
After an unsatisfying discussion with one of the maintainers of Skim.app, I have started looking around for a different solution and I have to admit that I haver overlooked following two options:

1. latex-suite has already taken care of this issue: https://github.com/vim-latex/vim-latex/blob/master/plugin/remoteOpen.vim#L4-L29  
   This could be a plugin on its own, however, adding a similar functionality to vimtex, would make the synctex feature more robust and self-contained.
2. A second option is following bash script: https://github.com/keflavich/macvim-skim/blob/master/macvim-skim/macvim-load-line#L14-L29
  This script 'macvim-load-line' requires the vim plugin 'WhichTab' available in the same repo.

I think the latex-suite solution would be the most convenient one for macvim-vimtex-skim.app user since they do not have to install anything to $PATH. On the other hand being able to open any file in the correct vim instance is a good reason to have something more generic like 'macvim-load-line'. Unfortunately the cli interface of 'macvim-load-line' is not compatible with vim/mvim which does not convince me to install.

What do you think? Is adding something like `remoteOpen.vim` to vimtex a feature you would consider?

**Comment by lervag** [comment]
Sorry for taking so long to answer. I don't really want to add this. It does not really have anything to do with LaTeX editing in itself, and could easily be a dedicated plugin. And that's exactly what I think it should be.

If you were to make a fork of this as a dedicated plugin, which I think you can do (but I guess you should check the license), then I would be very happy to mention it in the docs. I think it would be as simple as removing the latex-suite specific code (which is mostly debugging code), and to move the code into a dedicated plugin structure.

To summarize: There is nothing stopping you from using the latex-suite remoteLoad plugin, as it is fully standalone. If you only want to use it, you could remove the latex-suite code and bundle it as a dedicated plugin, and if so inclined, you could share it as a vim plugin.

**Comment by kiryph** [comment]
I already did what you suggested: I have added remoteLoad to my vim configuration.

I am not sure what I stick to in the long run. Derek Wyatt advocates in his video https://vimeo.com/4446112 to use only a single mvim/gvim instance. I have a question if I would follow this suggestion: Does vimtex support compilation of several maintex files in the same vim instance at the same time? Right now I do not use vimtex for compilation but `$ latexmk -pvc` in continuously update and preview mode in the terminal. So this is not an issue for me at the moment. However, I miss the integration of error log parsing. 
A second question would be: can I add a note to the statusline (configured possibly with airline/lightline) to see whether vimtex is in continuous mode for one or several maintex files. Something like this
`[vimtex-watching(x)]` `[vimtex-compiling(x/X)]` where x indicates that vimtex watches X files or is compiling x of X files.

**Comment by lervag** [comment]
Yes, vimtex supports multiple projects at the same time, where each may run a separate compilation process.

Vimtex provides a lot of the data very easily accessible. When a compilation is active, the PID is stored as `b:vimtex.pid`. Thus you could use this in a function for the statusline. I would propose something like `get(get(b:, 'vimtex', {}), 'pid', 0)`, which is a very safe way to get the PID number (will work both when the PID variable is not available (it should be available unless you disable the latexmk module), and when the `b:vimtex` data blob (as I call it) is not available (e.g. in non TeX buffers).

**Comment by nilsbecker** [comment]
i'm also wanting a way to address the right vim from skim for inverse search. is there a workable solution available at the moment?

**Comment by lervag** [comment]
I'm sorry, but I don't know. Perhaps/Hopefully, someone else might chime in and help...

**Comment by kiryph** [comment]
@nilsbecker: I think you should try out `https://raw.githubusercontent.com/vim-latex/vim-latex/master/plugin/remoteOpen.vim`

If you happen to use `dein.vim`, MacVim and Skim.app on macOS, the setup would look like following:

1. Add to your plugin list:
```
call dein#add(
		\ 'https://raw.githubusercontent.com/vim-latex/'
		\ . 'vim-latex/master/plugin/remoteOpen.vim',
		\ {'script_type' : 'plugin'} )
```
and run `:call dein#install()`
2. Configure Skim.app

![untitled](https://user-images.githubusercontent.com/371756/30423326-d03ab656-9942-11e7-98c4-2ef956f57f3a.png)

3. Test with two files in two different macvim windows.

**Comment by nilsbecker** [comment]
thanks! i'm using vundle at the moment. is the remoteOpen.vim script a full vim plugin that i can install as any other, or is it a single script extracted from vim-latex? -- i am guessing the latter since the plugin installation looks customized ? i guess i could see if vundle also allows doing something similar.

**Comment by nilsbecker** [comment]
hm, i found an option to specify a subdirectory of a full git repo as a plugin in vundle, but not an individual vimscript file. i guess i'll have to copy it locally. should not be a problem.

**Comment by nilsbecker** [comment]
ok, i got it working now like this:
1. clone vim-latex to my own github account
2. clone that locally
3. delete all files of vim-latex except plugin/remoteOpen.vim and the .git directory
4. git commit and push back to my github fork
5. in ~/.vimrc, add the line Plugin 'nilsbecker/vim-latex', {'name': 'remoteOpen'}
6. update plugins as usual in vundle.
7. enjoy backward search to multiple buffers across the same or different vim servers!

**Comment by kiryph** [comment]
Installation of the single file with `vundle` sounds quite complicated. I had a look into the doc of `vundle`. Apparently, it only supports installation via git (local or remote) which makes it impossible to grab only a single file from `vim-latex`. That's too bad.

I guess you are not interested in switching to [dein](https://github.com/Shougo/dein.vim). In this circumstance `dein` seems to be super convenient compared to vundle and possibly other plugin managers which are also restricted to `git clone`, such as`minpack` and `vim-plug`.

Anyhow, I am happy to help.
